More flexible --test option

mattias-p opened the pull request:
## Purpose

This PR implements flexible command line overriding of which test cases to run from the command line.

## Context

* This is an extension to the improvements in #333.

* Replaces #130. The point of that PR is to allow users to exclude test cases from the command line. This PR includes the same functionality using a different command line syntax.

* Leverages zonemaster/zonemaster-engine#1312 to properly handle the Basic test module (c.f. https://github.com/zonemaster/zonemaster-cli/pull/130#issuecomment-572948201).

* Handles the interaction of `--test` and `--profile` in accordance with https://github.com/zonemaster/zonemaster-cli/pull/130#issuecomment-573012219.

* For what it's worth, the idea for this design came to me in a discussion about https://github.com/zonemaster/zonemaster-engine/pull/1294.

## Changes

* The set of values accepted by `--test` is extended.

* Which test cases to include in the run is now controlled by setting the `test_cases` profile property and running Zonemaster::Engine->test_zone().

* When `--test=basic` or `--test=basic02` is specified, and basic02 considers the domain too broken to test, a CANNOT_CONTINUE message is now emitted, whereas it wasn't emitted before this PR.

## Reviewing

* Look especially for lacking test coverage.

## Testing

* Unit tests should pass.

matsduf commented:
>     * Before this PR CLI would call either `test_zone`, `test_module` or `test_method` in Zonemaster::Engine to run the test. After this PR CLI always calls `test_zone` and uses the `test_cases` profile property to control which tests are run. This means that the Basic tests are always included even when a single test case is specified.
> 
>     * Since the semantics of specifying a single test case or single test module is changed so that the Basic tests are included, an argument could be made that this is a breaking change, but I believe an argument could also be made that it's still just an addition.

That is a bad change. When working with test zones you want to be able to run just the specific test case and nothing else. 

Secondly, with that change, if Basic fails completely no other test case is run, and that is not what we want when testing and creating test zones.

I am against a change that makes it impossible to run just the selected test case.

mattias-p commented:
All I'm hearing is "no" and "bad". Is there anything about this proposal that you do like?

Also, I'm trying to understand your objection here. AFAICT you're saying that when a user requests that a single test case is run on a domain that doesn't fulfill the requirements posed by Basic, the requested test case still needs to run. Is that correct? And if so, could you explain why this is important, because it's not obvious to me.

matsduf commented:
> All I'm hearing is "no" and "bad". Is there anything about this proposal that you do like?

It would be good to be able to exclude test cases without creating a new profile. It would be good to be able to select two or more test cases and they are run without overlap.

Sometimes it is probably good that Basic is run before a selected test.

> Also, I'm trying to understand your objection here. AFAICT you're saying that when a user requests that a single test case is run on a domain that doesn't fulfill the requirements posed by Basic, the requested test case still needs to run. Is that correct? And if so, could you explain why this is important, because it's not obvious to me.

I want to see what happens in that test case for the test zone even when there are errors. And I want minimal output without the extra output from Basic when testing test zones. Sometimes I have to run with `--level debug` which gives many lines, which will be anymore if Basic is run.

Give the possibility to run a test without adding Basic.

If Basic is not in the profile, why should Basic still be run? The magic could rather be to put the Basic test cases in the profile first, if there are any.

mattias-p commented:
> I want to see what happens in that test case for the test zone even when there are errors. And I want minimal output without the extra output from Basic when testing test zones. Sometimes I have to run with --level debug which gives many lines, which will be anymore if Basic is run.

> If Basic is not in the profile, why should Basic still be run? The magic could rather be to put the Basic test cases in the profile first, if there are any.

Ok. Yeah, I see how this could be very valuable in development contexts. And I also don't particularly like that Basic is forced.

I have an idea of how we could simplify things in Engine in a way that supports this. I think I'll make a draft PR there too and we can let this PR rest until Engine is in a better shape to support it.

mattias-p committed:
Expression language for specifying tests to run

matsduf commented:
@mattias-p, is this still relevant or can it be closed now?

mattias-p commented:
This depends on #371. I plan to rebase this PR the documentation improvements after #371 PR have been merged. It seems somewhat unreasonable to get this merged for 2024.2 so I postponed it.

mattias-p commented:
I rebased this onto latest develop and fixed some typos added some unit tests. I also tightened some validations for sanity reasons, but that shouldn't have any effect on zonemaster-cli behavior.

matsduf commented:
This is not true anymore. Please update description.

>     * This PR contains the commits from [Organize --help text #389](https://github.com/zonemaster/zonemaster-cli/pull/389), which should be merged before this one.

matsduf commented:
From `man zonemaster-cli`:

```
   Testing Options
       --test=TESTCASE, --test=TESTMODULE
           Limit the testing suite to run only the specified tests.  (default:
           the empty string)

           EXPR is an expression specifying or modifying the set of tests to
           run.  Here are some examples:
```
Shouldn't it be `--test=EXPR`?

matsduf commented:
In the man page I find the following two examples,

```
           --test=all-nameserver+nameserver03
               runs all test cases of all test modules except for the
               Nameserver test module in which only the Nameserver03 test case
               is run.

           --test=-dnssec
               removes all test cases of the DNSSEC module from the set of
               tests that would otherwise have run.
```

Does `--test=-dnssec` and `--test=all-dnssec` mean the same thing?

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<nameserver>
+
+runs the entire Nameserver test module, and nothing else.
```
```suggestion
runs the entire Nameserver test module, including any test cases
disabled by the profile, but nothing else.
```

mattias-p replied:
I agree that the documentation should make this clear, but I don't think it needs to be repeated for every example. I'll add a separate paragraph clarifying how it interacts with other options.

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+EXPR is an expression specifying or modifying the set of tests to run.
+Here are some examples:
+
+=over 4
```
```suggestion
=over 4

=item --test=all

runs all defined test cases including any test cases 
disabled by the profile.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<dnssec-dnssec05>
+
+runs the entire DNSSEC test module except the DNSSEC05 test case, and nothing
+else.
```
```suggestion
runs the entire DNSSEC test module including any test cases
disabled by the profile, except the DNSSEC05 test case,
and nothing else.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=
+
+(i.e., the empty string) leaves the set of tests to run unmodified.
```
```suggestion
(i.e., the empty string) leaves the set of tests to run unmodified,
which means that all test cases enabled by the profile will
be run.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
 =item B<--test>=TESTCASE, B<--test>=TESTMODULE
```
```suggestion
=item B<--test>=TESTCASE, B<--test>=TESTMODULE, B<--test>=EXPR
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+=item --test=B<-dnssec>
+
+removes all test cases of the DNSSEC module from the set of tests that would
+otherwise have run.
```
```suggestion
otherwise have run, i.e. the test cases enabled by the profile.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+=item --test=B<+syntax01+syntax02>
+
+adds the Syntax01 and Syntax02 test cases to the set of tests that would
+otherwise have run.
```
```suggestion
otherwise have run, i.e. the test cases enabled by the profile.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -114,14 +114,50 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<all-nameserver+nameserver03>
+
+runs all test cases of all test modules except for the Nameserver test module in
+which only the Nameserver03 test case is run.
```
```suggestion
runs all defined test cases of all defined test modules (disregardning the profile)
except for the Nameserver test module in which only the Nameserver03 test
case is run.
```

matsduf commented:
Nice to have it consistent. Nice to be able to control what test cases to run without creating a custom profile.

mattias-p commented:
> Shouldn't it be --test=EXPR?

Fixing.

> Do `--test=-dnssec` and `--test=all-dnssec` mean the same thing?
> 
> ---> No
> 
> Do `--test=` and `--test=all` mean the same thing?
> 
> ---> No

Agreed. In case you modify the `test_cases` property using --profile they won't be the same.

mattias-p commented:
I think the current implementation is flawed regarding the interpretation of this command sequence:
```sh
printf "%s\n" "--test=-delegation01" > ~/.zonemaster/cli.args
zonemaster-cli example.com
zonemaster-cli --test=-delegation02 example.com
```

With the current proposal, in the first command delegation01 would be disabled, but in the second command it would be enabled again. This is because the second command would effectively be `zonemaster-cli --test=-delegation01 --test=-delegation02 example.com` and the second --test would override the first one. It would make much more sense if both --test arguments were applied as they appear on the command line. `zonemaster-cli --test=-delegation01 --test=basic01` would still mean execute basic01 and nothing else.

matsduf commented:
> I think the current implementation is flawed regarding the interpretation of this command sequence:
> 
> ```shell
> printf "%s\n" "--test=-delegation01" > ~/.zonemaster/cli.args
> zonemaster-cli example.com
> zonemaster-cli --test=-delegation02 example.com
> ```
> 
> With the current proposal, in the first command delegation01 would be disabled, but in the second command it would be enabled again. This is because the second command would effectively be `zonemaster-cli --test=-delegation01 --test=-delegation02 example.com` and the second --test would override the first one. It would make much more sense if both --test arguments were applied as they appear on the command line. `zonemaster-cli --test=-delegation01 --test=basic01` would still mean execute basic01 and nothing else.

I think it is correct that `--test` on the command line overrides `--test` in the configuration file. That is how it works in other cases.

mattias-p commented:
We have two kinds of option today: overwriting ones and aggregate ones. Most of our options overwrite, but --test, --ns and --ds aggregate. I'm arguing that it makes more sense to keep --test as an aggregating option to change it into an overwriting one. I.e., `--test=basic01 --test=+basic02` should be equivalent to `--test=basic01+basic02`, not to `--test=+basic02`.

mattias-p commented:
I pushed an update makes --test aggregate again. It's in a separate commit in case you want to play with both the overwriting version and the aggregating one. The man page needs an update to clarify the semantics in of repeated --test options but I haven't got around to that.

I haven't forgotten your unresolved review comments, @matsduf. It's just that before I take care of them I want to settle this overwriting/aggregating business.

matsduf commented:
> We have two kinds of option today: overwriting ones and aggregating ones. Most of our options overwrite, but --test, --ns and --ds aggregate. I'm arguing that it makes more sense to keep --test as an aggregating option than to change it into an overwriting one. I.e., `--test=basic01 --test=+basic02` should be equivalent to `--test=basic01+basic02`, not to `--test=+basic02`.

There is a clear difference between `--ns` and `--ds`, on one hand, and  `--test`, on the other hand.

Options `--ns` and `--ds` are typically used for a specific zone. It is unlikely that you would put that into the the configuration file `cli.args`. Therefor it works to have them aggregating. On the other hand, it would also work to change them so that only the last one is used, but so that you can specify multiple NS or DS, respectively, for the option, e.g. with a `+` sign or in quotemarks.

Option `--test` can be used in `cli.args` instead of creating a custom profile. Then it should be possible to override that with the settings on the command line.

mattias-p committed:
Tidy

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+=over 4
+
+=item The full set of all test methods.
```
All test methods in the default profile?

mattias-p replied:
Good idea. Though I'll express it in another way.

mattias-p replied:
On second thought, no. The TestCaseSet doesn't have any concept of profiles at all. It just works with whatever you tell it in the constructor.

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+=head1 NAME
+
+    Zonemaster::CLI::TestCaseSet - A mutable set of test methods names.
```
It is confusing when it says "test methods" when else in Zonemaster refer to "test cases".

```suggestion
    Zonemaster::CLI::TestCaseSet - A mutable set of test method names (test case names).
```

mattias-p replied:
I agree. I'll just avoid referring to test methods altogether.

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+A TestCaseSet primarily represents an immutable full set of test methods and a
+mutable subset thereof. The full set of test methods is distributed across the
+set of test modules.
```
```suggestion
A TestCaseSet primarily represents an immutable full set of test methods (test cases) and a
mutable subset thereof. The full set of test methods (test cases) is distributed across the
set of test modules.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,65 @@ Print the effective profile used in JSON format, then exit.
+An EXPR starting with a term (e.g., C<all>, C<nameserver01>) completely replaces
+the current value of C<test_cases>.
+An EXPR starting with C<+> or C<-> modifies the existing C<test_cases>.
+An empty EXPR (C<--test=>) leaves C<test_cases> unchanged.
```
Can you clearify what it means to "leave 'test_cases' unchanged'?

mattias-p replied:
I changed how I refer to test_cases.

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,65 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=
+
+leaves the existing C<test_cases> unmodified.
```
See above.

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,65 @@ Print the effective profile used in JSON format, then exit.
+Multiple C<--test> options may be provided; they are applied sequentially.
+When used with C<--profile>, that option is applied first.
+
+Use C<--list-tests> to list all available test cases and test modules.
```
```suggestion
Use C<--list-tests> to list all available test cases and test modules in the default profile.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,65 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<all>
+
+selects all available test cases.
```
```suggestion
selects all available test cases in the default profile.
```

mattias-p replied:
Fixed.

matsduf reviewed (changes requested):
I have tested and it seems to work as I think it was meant to work.

mattias-p committed:
Updates after review comments

mattias-p committed:
Update script/zonemaster-cli

Co-authored-by: Mats Dufberg <mats.dufberg@iis.se>

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+=over 4
+
+=item The full set of test cases.
```
```suggestion
=item The full set of test cases in the default profile.
```

mattias-p replied:
As I replied before, the TestCaseSet doesn't have any concept of profiles at all. It just works with whatever you tell it in the constructor.

matsduf replied:
When you select "all" you get all test cases in the default profile.

Could you give a better explanation and description of `TestCaseSet.pm`? I do not understand.

mattias-p replied:
> When you select "all" you get all test cases in the default profile.

This true if and only if you provide "all test cases in the default profile" to the TestCaseSet constructor. In other words, it depends on how the module is used. The documentation of a module should not depend on how that module is being used.

matsduf replied:
Can you please give a more explicit description of the module? I still do not understand it.

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,66 @@ Print the effective profile used in JSON format, then exit.
+An EXPR starting with a term (e.g., C<all>, C<nameserver01>) completely replaces
+the selection (i.e., the value of the C<test_cases> profile property).
+An EXPR starting with an operator applies a delta to the selection.
+An empty EXPR (C<--test=>) leaves the selection unchanged (formally, it is
+updated by applying the identity transformation).
```
What does "leaves the selection unchanged" mean? For what can I use that?

mattias-p replied:
`--test=` is useful the same way `x + 0` is useful. I don't expect you to type it out on the command line, but it could make life easier if you want to call zonemaster-cli programmatically.

matsduf replied:
Can you please clarify that, preferably below when it listed on its own line.

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,66 @@ Print the effective profile used in JSON format, then exit.
+
+adds the Syntax01 and Syntax02 test cases to the selection.
+
+=item --test=
+
+leaves the selection unchanged.
```
When do I want to use that feature? What does it give me?

matsduf replied:
Can you please clarify here that it will not do anything?

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+=over 4
+
+=item The full set of test cases.
```
```suggestion
=item The full set of test cases as provided to the TestCaseSet constructor for the current object.
```

matsduf commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,233 @@
+
+=item The singleton set of a single test cases
+
+Terms matching the name of a test case or the concatenation of a test module,
+a slash and a test cases belonging to that test module.
```
```suggestion
Terms matching the name of a test case (e.g. "Case10") or the concatenation of a test module,
a slash and a test cases belonging to that test module (e.g. "Case/Case10").
```

mattias-p committed:
Update lib/Zonemaster/CLI/TestCaseSet.pm

Co-authored-by: Mats Dufberg <mats.dufberg@iis.se>

mattias-p committed:
Update lib/Zonemaster/CLI/TestCaseSet.pm

Co-authored-by: Mats Dufberg <mats.dufberg@iis.se>

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,66 @@ Print the effective profile used in JSON format, then exit.
+
+adds the Syntax01 and Syntax02 test cases to the selection.
+
+=item --test=
+
+leaves the selection unchanged.
```
```suggestion
=item --test=

empty value is valid, but does not change the selection of test cases
```

mattias-p committed:
Reworked documentation and naming in TestCaseSet

mattias-p commented:
> Can you please give a more explicit description of the module? I still do not understand it.

I read through the module documentation again and I agree there was some room for improvement. So I worked through it again.

mattias-p committed:
Further tweaks to documentation

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,67 @@ Print the effective profile used in JSON format, then exit.
+An empty EXPR (C<--test=>) keeps the current selection unchanged (i.e., it is a no-op).
+
+Multiple C<--test> options may be specified; they are applied in the order they are given.
+When used together with the C<--profile>, that option is applied first.
```
To make it clearly unambiguous:

```suggestion
When used together with the C<--profile>, the profile option is applied first.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,67 @@ Print the effective profile used in JSON format, then exit.
+An EXPR starting with a term (e.g., C<all>, C<nameserver01>) I<replaces>
+the current selection.
+An EXPR starting with an operator (C<+> or C<->) I<modifies> the current selection.
+An empty EXPR (C<--test=>) keeps the current selection unchanged (i.e., it is a no-op).
```
```suggestion
An empty EXPR (C<--test=>) does not modify the selection (i.e., it is a no-op).
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,67 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<all>
+
+selects all test cases in the default profile.
```
```suggestion
selects all test cases in the default profile, i.e. overrides any custom profile.
```

matsduf commented on script/zonemaster-cli:
```diff
@@ -111,17 +108,67 @@ Print the effective profile used in JSON format, then exit.
+
+=item --test=B<nameserver>
+
+selects all test cases in the Nameserver test module.
```
```suggestion
selects all test cases in the Nameserver test module in the default profile.
```

matsduf reviewed (commented):
This fine, but I have some comments on the documentation to make is clearer. The functions work well.

mattias-p committed:
Apply suggestions from code review

Co-authored-by: Mats Dufberg <mats.dufberg@iis.se>

mattias-p committed:
Update script/zonemaster-cli

Co-authored-by: Mats Dufberg <mats.dufberg@iis.se>

marc-vanderwal commented on lib/Zonemaster/CLI/TestCaseSet.pm:
```diff
@@ -0,0 +1,271 @@
+        }
+    }
+    else {
+        croak "Unrecognized operator '$op'";
```
This line is the only thing not covered by the existing unit tests.

marc-vanderwal commented on t/test_case_set.t:
```diff
@@ -0,0 +1,303 @@
+        } ## end for my $case ( @cases )
+    };    ## end 'new' => sub
+
+    subtest 'apply_modifier' => sub {
```
There should be a test case checking whether there is a `croak` if passing an illegal modifier.

marc-vanderwal reviewed (commented):
Looks fine.

I’ve run the test suite through Devel::Cover and found a single line in `TestCaseSet.pm` that wasn’t covered by the test suite. Otherwise it’s 100% covered.

mattias-p committed:
Add test for invalid operator

mattias-p commented:
I've rebased and added a unit test for invalid operators. The unit test is slightly messy, but I didn't know how to do it more cleanly. Please review again.
